src/runmode-dpdk.c

```diff
@@ -0,0 +1,1274 @@
+        rxq_conf.rx_free_thresh = 0;
+        rxq_conf.rx_drop_en = 0;
+        SCLogInfo(
+                "Creating Q %d of P %d using desc RX: %d TX: %d RX htresh: %d RX pthresh %d wtresh "
+                "%d free_tresh %d drop_en %d Offloads %lu",
+                queue_id, iconf->port_id, iconf->nb_rx_desc, iconf->nb_tx_desc,
+                rxq_conf.rx_thresh.hthresh, rxq_conf.rx_thresh.pthresh, rxq_conf.rx_thresh.wthresh,
+                rxq_conf.rx_free_thresh, rxq_conf.rx_drop_en, rxq_conf.offloads);
+
+        retval = rte_eth_rx_queue_setup(iconf->port_id, queue_id, iconf->nb_rx_desc,
+                iconf->socket_id, &rxq_conf, iconf->pkt_mempool_array[queue_id]);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT,
+                    "Error (err=%d) during initialization of device queue %u of port %u", retval,
+                    queue_id, iconf->port_id);
+            SCReturnInt(retval);
+        }
+    }
+
+    for (uint16_t queue_id = 0; queue_id < iconf->nb_rx_queues; queue_id++) {
+        txq_conf = dev_info->default_txconf;
+        txq_conf.offloads = port_conf->txmode.offloads;
+        retval = rte_eth_tx_queue_setup(
+                iconf->port_id, queue_id, iconf->nb_tx_desc, iconf->socket_id, &txq_conf);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT,
+                    "Error (err=%d) during initialization of device queue %u of port %u", retval,
+                    queue_id, iconf->port_id);
+            SCReturnInt(retval);
+        }
+    }
+
+    SCReturnInt(0);
+}
+
+int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
+{
+    int retval;
+
+    if (iconf->out_iface != NULL) {
+        retval = rte_eth_dev_get_port_by_name(iconf->out_iface, &iconf->out_port_id);
+        if (retval != 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) during obtaining port id of %s", retval,
+                    iconf->out_iface);
+            SCReturnInt(retval);
+        }
+
+        if (rte_eth_dev_socket_id(iconf->port_id) != rte_eth_dev_socket_id(iconf->out_port_id)) {
+            SCLogWarning(SC_WARN_DPDK_CONF, "%s and %s are not on the same NUMA node", iconf->iface,
+                    iconf->out_iface);
+        }
+    }
+    SCReturnInt(0);
+}
+
+int DeviceConfigure(DPDKIfaceConfig *iconf)
+{
+    // configure device
+    int retval;
+    struct rte_eth_dev_info dev_info;
+    struct rte_eth_conf port_conf;
+
+    retval = rte_eth_dev_get_port_by_name(iconf->iface, &(iconf->port_id));
+    if (retval < 0) {
+        SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) when getting port id of %s Is device enabled?",
+                retval, iconf->iface);
+        SCReturnInt(retval);
+    }
+
+    if (!rte_eth_dev_is_valid_port(iconf->port_id)) {
+        SCLogError(SC_ERR_DPDK_INIT, "Specified port %d is invalid", iconf->port_id);
+        SCReturnInt(retval);
+    }
+
+    retval = rte_eth_dev_socket_id(iconf->port_id);
+    if (retval < 0) {
```

Comment:
**this comparison is incorrect for single NUMA and Virtual machine**. rework.

Reply:
I have already called this out since for single physical NUMA this API return non zero value also.

Current DPDK support is focused to implement physical machines and physical NICs.
[VV] DPDK has limitation to run on physical virtual or docker. As far I have reviewed there is no issue in your current code which does not allow PCIe pass though for guest OS.

rework the code to support vendor neutral code.

Reply:
> I have already called this out since for single physical NUMA this API return non zero value also.

https://doc.dpdk.org/api/rte__ethdev_8h.html#ad032e25f712e6ffeb0c19eab1ec1fd2e This is not mentioned in the documentation. Are you sure? 

> As far I have reviewed there is no issue in your current code which does not allow PCIe pass though for guest OS.

With my limited time, I am afraid that the proper implementation of virtual environment/docker will need to wait for another pull request.